common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+	return &limiter[T]{
+		options: options,
+	}
+}
+
+// enqueue a task in the limiter's buffer.
+func (lim *limiter[T]) enqueue(task T) {
+	lim.buffer = append(lim.buffer, task)
+}
+
+// dequeue a task from the limiter's buffer.
+func (lim *limiter[T]) dequeue() (task T, ok bool) {
+	if len(lim.buffer) > 0 {
+		task, lim.buffer = lim.buffer[0], lim.buffer[1:]
+		return task, true
+	}
+	return task, false
+}
+
+// reserve a concurrency slot. Returns true on success.
+func (lim *limiter[T]) reserve() bool {
+	if lim.reservations < lim.options.Concurrency() {
+		lim.reservations++
+		return true
+	}
+	return false
+}
+
+// release a reservation.
+func (lim *limiter[T]) release() {
+	lim.reservations--
+}
+
+// MultiDestinationSchedulerOptions are options for creating a [MultiDestinationScheduler].
+type MultiDestinationSchedulerOptions[K comparable, T Task] struct {
+	// A function to determine the destination of a task.
+	KeyFn func(T) K
+	// Limiter options for a given destination.
+	LimiterOptionsFn func(K) LimiterOptions
+	Logger           log.Logger
+}
+
+// MultDestinationScheduler for limiting concurrency and RPS (and later applying a circuit breaker) on a per destination
+// basis. A task's destination is determined based on a provided key function.
+// Accepts and buffers tasks that exceed the destination's concurrency limit to be picked up as soon as there's more
+// availability.
+type MultiDestinationScheduler[K comparable, T Task] struct {
+	options      MultiDestinationSchedulerOptions[K, T]
+	waitGroup    sync.WaitGroup
+	stopCtx      context.Context
+	stopFn       context.CancelFunc
+	limitersLock sync.RWMutex
+	limiters     map[K]*limiter[T]
+}
+
+func NewMultiDestinationScheduler[K comparable, T Task](options MultiDestinationSchedulerOptions[K, T]) *MultiDestinationScheduler[K, T] {
+	stopCtx, stopFn := context.WithCancel(context.Background())
+	return &MultiDestinationScheduler[K, T]{
+		options:  options,
+		stopCtx:  stopCtx,
+		stopFn:   stopFn,
+		limiters: make(map[K]*limiter[T]),
+	}
+}
+
+func (*MultiDestinationScheduler[K, T]) Start() {
+	// noop
+}
+
+// Stop signals running tasks to stop, aborts any pending tasks and waits up to a minute for all running tasks to
+// complete.
+func (s *MultiDestinationScheduler[K, T]) Stop() {
+	s.stopFn()
+	s.abortBufferedTasks()
+	if success := common.AwaitWaitGroup(&s.waitGroup, time.Minute); !success {
+		s.options.Logger.Warn("MultiDestinationScheduler timed out waiting for tasks to complete")
+	} else {
+		s.options.Logger.Debug("MultiDestinationScheduler drained")
+	}
+}
+
+func (s *MultiDestinationScheduler[K, T]) abortBufferedTasks() {
+	s.limitersLock.Lock()
+	defer s.limitersLock.Unlock()
+
+	for _, lim := range s.limiters {
+		lim.Lock()
+		for _, task := range lim.buffer {
+			task.Abort()
+		}
+		lim.buffer = nil
+		lim.Unlock()
+	}
+}
+
+// Submit is not implemented as it is never invoked.
+func (*MultiDestinationScheduler[K, T]) Submit(T) {
+	panic("unimplemented")
```

Comment:
This is a bit surprising. Why is it required by the interface if it's never invoked? Any way we can better design the types to avoid this? If it's not immediately clear whether this is fixable, would you mind making an issue for it? I think this is actually something that would be appropriate for a public GH issue as I could see an open source contributor solving this.

Reply:
We might be able to get rid of the common.Scheduler interface. Queue processing logic for transfer/timer, etc. can define an interface with only the TrySubmit() method and replication queue can define its own interface with only the Submit() method. It might be hard to Scheduler wrappers though...